### DISCUSSION TITLE
C# Plugins

### DISCUSSION THREAD

**Discussion by LennardF1989** [praydog/REFramework#1018] [General Questions]
Not seen a topic for this yet, but looking at how the commits for this C#-branch are about 7 hours old, I reckon none has actually played with it yet. So let me be the first!

At first, I want to express a huge thank you to @praydog . Not just for the C# support, but especially when it comes to the project itself. When I encounter open-source C++ projects I'm always dreading the thought of having to compile them locally, because it's usually an absolute pain to do on Windows. But this project, man, it even has a batch file that compiles it using VS2022 and does all the work for me. And it even worked to just copy it to the csharp-api folder (with some slight modification). I absolutely love it, thank you! (Only had some issue with an old CMake version, but that was my own bad).

That said, gonna test this out and will report weird things I encounter :)

**Comment by praydog** [comment]
Be sure to read the preliminary docs here https://cursey.github.io/reframework-book/api_cs/general/index.html

This is rapidly changing over time so be aware, another reason why it's not fully merged.

**Comment by LennardF1989** [comment]
Yea, that's what triggered me to check it out!

But it seems it fails to load REFramework.NET.dll ingame with a self-compiled dinput8.dll for DD2. Error-log is not really helpful as to why it won't load.

What game are you using to test things? Guess I'll just hook up a debugger and try to see where it fails.

**Comment by praydog** [comment]
Use the symlink script that's in the csharp-api folder to set up your game directory for development. REFramework.NET.dll depends on the ijwhost.dll that came with it, and also a few other DLLs that will need to be placed in plugins/managed/dependencies

Using the python script should speed things along.

**Comment by praydog** [comment]
Plugins folder
![top](https://github.com/praydog/REFramework/assets/2909949/a9a60a9d-0993-4229-a288-6b4106e6afa2)

Dependencies folder
![des](https://github.com/praydog/REFramework/assets/2909949/4f484efc-2910-4e23-9f55-f27db7e76a8c)

**Reply by LennardF1989** [reply]
Ah, didn't see the .py script - thanks! I got it working now :) Let's see how far I can push this thing :P

**Reply by LennardF1989** [reply]
How would you like to go about reporting issues since it's rapidly changing? In this Discussion, or normal Issues?

**Reply by praydog** [reply]
I was considering making a separate repo for REFramework.NET considering it's a full blown plugin but not sure yet.

I'm fine with issues for now.

**Comment by LennardF1989** [comment]
I'll use this post for some findings before reporting them:
- The main thread of the plugin cannot be blocking, otherwise it will simply block the game, too.
- ~~To get around that, I use a Task/Thread, but it looks like that will cause DD2 to start lagging a few minutes into the main menu.~~
- Would be nice to skip the DLL generation once it's done, as it causes quit a major startup time.
- The following error is always in the console: 
```
Unloading dynamic assemblies...
Object reference not set to an instance of an object.
   at System.Runtime.Loader.AssemblyLoadContext.Unload()
   at REFrameworkNET.PluginManager.UnloadDynamicAssemblies()
   at REFrameworkNET.PluginManager.LoadPlugins(UInt64 param_raw)
```
- Has a tendency to hang on startup or outright crash. Loading the .NET Runtime seems quite hefty. Can cause the Steam overlay from not being properly injected. My eyes are "bad", so I like to set font size to 24, but this can cause the Steam overlay to get scaled as well and break its input.

~~My first thought of the error was that this was causing the slowdown, because it may be destroying the plugin while it's not actually done. But removing the thread completely stops the slowdowns.~~

**Reply by praydog** [reply]
Generally I would recommend you do not create your own threads and call any game functions within them. This may cause some unknown behavior. This recommendation holds within C++ plugins as well. You should use the REFrameworkNET.Callbacks system and do anything that calls into the game with those. The Main function for plugins is also safe to call most game logic from.

If you need ACTUAL threads, try to only do independent logic within them that does not ever call any class methods on a game object, or any SDK functionality at all generally.

Also the dynamic assemblies immediately unloading after being finished is a test thing. My goal here is to have it similar to Lua where the user can just choose to reload the scripts at will.

**Reply by LennardF1989** [reply]
Yea, what I'm trying to do is get a simple socket connection going so I can communicate with the game from an external application (and other way around). I must have been doing something weird, because after restructuring the code it works as intended now.

Really like where it is now, all this needs is imgui exposed functions :)

This does raise some concerns with security though. The Lua is somewhat sandboxed, this is next-level system access... On the other hand, this is not really your concern.

**Reply by praydog** [reply]
C++ plugins offer the same level of access. Should I be sandboxing it to certain parts of the .NET ecosystem? I don't know. The C# API has the potential to be very powerful, so restricting it too much feels like a waste. This is also my first real deep dive into .NET after making a handful of hobbyist C# apps, so I'm not sure what the standard is here generally in the .NET community.

**Reply by LennardF1989** [reply]
> I'm not sure what the standard is here generally in the .NET community

I've never seen a C# API layer for game mods that was in any way constrained/sandboxed - I wouldn't bother. The reason I said it out loud was because I briefly forgot that C++ plugins exist with the same security concerns. In the end, the user is responsible for picking the right mods and not just trust everything.

**Reply by praydog** [reply]
I believe the game "S&box" has an actual sandboxed C# scripting API. There are some remarks about what they restrict here: https://developer.valvesoftware.com/wiki/S%26box

So just something to keep in mind for later.

**Reply by praydog** [reply]
csharp-api branch has ImGui.NET support now. Requires a full recompile of REFramework itself. Redirects cimgui to dinput8.dll which acts as cimgui. Still trying to see if there's a less hacky way of doing that. Bunch of other stuff added too.

**Reply by LennardF1989** [reply]
I saw the commits yesterday, awesome stuff. I'll give it some testing throughout this week :)

**Reply by LennardF1989** [reply]
Played around a bit with the ImGui integration and seems to work fine! Whole C# integration feels a lot more stable overall. Things I've noticed:
- ImGui does not seem to be respecting (all) `PushStyleVar` - particularly I try to use `WindowPadding` and `WindowBg`, but it doesn't seem to have any effect. `WindowRounding` on the other hand does work, for example.
- `OnUnload` seems to crash the game if you try to access a resource (eg. Disposing it) that was created as part of `PluginEntryPoint`. I'm using a TcpListener and if I don't use `OnUnload`, it will just remain actively listening for connections (on a new thread), even after unloading the script. But I also don't get a chance to properly clean it up on `OnUnload` now. Can't seem to get a proper stacktrace of this. Workaround is probably signaling the thread to kill itself and close the server naturally like that.
- `ImGui.GetMainViewport()` returns 0 for X for me. I have a 2560x1440 screen, and it will return a Vector2(0,1440). Actually, looks like X/Y might even be swapped, so it puts 1440 in X.
